### PR TITLE
Add --no-listen option to disable listening for incoming connections

### LINKED ISSUE #1025: Option to disallow inbound channels
This would be useful for decommissioning a node gracefully. I don't want new nodes to connect to me while I'm waiting for `ONCHAIND_OUR_UNILATERAL`'s to finish. I suppose I could simply not set `--ipaddress` but that seems a bit of a hack.

### PR DESCRIPTION
**Pull request by cdecker**
Fixes #1025

### CODE REVIEW

**Comment by rustyrussell** [comment]
NAK, use `--port=0`.  We could make it an alias, of course, but that's a much smaller patch I think!

**Comment by practicalswift** [comment]
If we go the port 0 route, then please make an alias: `--no-listen` is more user-friendly than `--port=0`.

**Comment by cdecker** [comment]
Ok, added `--no-listen` as an alias to `--port=0` :-)

**Comment by ZmnSCPxj** [comment]
While the linked issue is indeed about disabling listening to incoming connections, I wonder if this truly solves the issue (decommissioning a node, and thus disabling incoming *channel* opens while we wait for unilateral closes to finish).

Should we also not add an option to immediately error any `open_channel` messages?

After all, presumably when started with `--no-listen` there exist current channels, `gossipd` autoconnects to peers with open channels, so there are outgoing connections.  But we can still receive `open_channel` messages from those outgoing connections; it might be possible for one connection to successfully close bilaterally their channel, then autopilot (or other automated script) on that node selects us to make a channel to after the close finishes, using the outgoing connection to send the `open_channel` message.

Or do I misunderstand something.

**Comment by cdecker** [comment]
We already have the `--dev-no-reconnect` option which will suppress any outgoing connections as well, so with the combination of the two we have isolated our node. Arguably we could create a decommission-mode that starts the node in a safe configuration, closes all channels and sweeps funds to an outside wallet, but I'd rather have that be driven by an outside script.

**Comment by ZmnSCPxj** [comment]
@cdecker but I suppose that the `dev` there implies it exists only if compiled with `DEVELOPER=1`, unless we remove that requirement?

**Comment by cdecker** [comment]
That's correct, we could make it a non-dev though, since it's rather useful.

**Comment by ZmnSCPxj** [comment]
However during decommissioning, it would be nice (not an absolute requirement) if it would connect to all peers and at least attempt a bilateral close. If all peers successfully close bilaterally all channels, then the time to decommission a node is shortened (no need to wait for the unilateral close finish).  But I suppose, not so high priority.

**Comment by rustyrussell** [comment]
OK, let's drop --no-listen and do --offline, which does both?  Never connects out, doesn't allow connections in?  Maybe later we'd allow manual connections in such a state, but for the moment, don't even allow that?

**Comment by cdecker** [comment]
I'd say `--offline` will not automatically try to connect, but manually connecting using the JSON-RPC `connect` call will still work. That should cover @ZmnSCPxj collaborative decommission I think. We can drive the actual decommission from a script.

**Comment by cdecker** [comment]
Ok, now it's called `--offline` and also disables auto-reconnect, but still allows manual connect.

Ping @ZmnSCPxj and @rustyrussell

File: gossipd/gossip.c [hunk 1]
```diff
@@ -1509,7 +1509,10 @@ static struct io_plan *gossip_init(struct daemon_conn *master,
 	}
 	daemon->rstate = new_routing_state(daemon, &chain_hash, &daemon->id);
 
-	setup_listeners(daemon, port);
+	if (port)
+		setup_listeners(daemon, port);
+	else
+		status_trace("Not listeing for incoming connections due to --no-listen");
```
Comment (rustyrussell) [inline gossipd/gossip.c]: s/listeing/listening/ and /--no-listen/--offline/.

Also, this duplicates the check inside setup_listeners; one should go...

Comment (cdecker) [reply gossipd/gossip.c]: Will remove the outer check.

Comment (cdecker) [reply gossipd/gossip.c]: Fewer chances for typos :wink:

**Comment by cdecker** [comment]
@rustyrussell it doesn't understand review-comments just yet, will fix that tomorrow.

**Pull request merged**